sklearn/cluster/dbscan_.py

```diff
@@ -118,18 +125,30 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     if metric == 'precomputed':
         D = pairwise_distances(X, metric=metric)
         neighborhoods = np.empty(X.shape[0], dtype=object)
-        neighborhoods[:] = [np.where(x <= eps)[0] for x in D]
+        if sparse.issparse(D):
+            D_mask = D.data <= eps
+            masked_indices = D.indices.astype(np.intp)[D_mask]
+            masked_indptr = np.cumsum(D_mask)[D.indptr[1:] - 1]
+            # insert the diagonal
```

Comment:
Maybe you could expand a bit that comment because it's not trivial why the zero-valued element of the diagonal need to be materialized in the distance matrix.